File: uilib/src/main/java/dev/entao/ui/util/RoundBackgroundSpan.java
package dev.entao.ui.util;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.text.style.ReplacementSpan;
import dev.entao.appbase.App;

public class RoundBackgroundSpan extends ReplacementSpan {
	private int backColor = Color.GRAY;
	private int textColor = Color.argb(255, 30, 30, 30);
	private int radius = App.INSTANCE.dp2px(4);

	public RoundBackgroundSpan() {

	}

	public RoundBackgroundSpan(int backColor, int textColor, int radiusDp) {
		setBackColor(backColor);
		setTextColor(textColor);
		setRadius(radiusDp);
	}

	public void setBackColor(int backColor) {
		this.backColor = backColor;
	}

	public void setTextColor(int textColor) {
		this.textColor = textColor;
	}

	public void setRadius(int dp) {
		this.radius = App.INSTANCE.dp2px(dp);
	}

	public void draw(Canvas canvas, CharSequence text, int start, int end, float x, int top, int y,
			int bottom, Paint paint) {
		RectF rect = new RectF(x, top, x + measureText(paint, text, start, end), bottom);
		paint.setColor(backColor);
		canvas.drawRoundRect(rect, radius, radius, paint);
		paint.setColor(textColor);
		canvas.drawText(text, start, end, x, y, paint);
	}

	public int getSize(Paint paint, CharSequence text, int start, int end, Paint.FontMetricsInt fm) {
		return Math.round(measureText(paint, text, start, end));
	}

	private float measureText(Paint paint, CharSequence text, int start, int end) {
		return paint.measureText(text, start, end);
	}
}

File: uilib/src/main/java/dev/entao/ui/util/RoundBackgroundSpan.kt
package dev.entao.ui.util

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.text.style.ReplacementSpan
import dev.entao.kan.appbase.App

class RoundBackgroundSpan(backColor: Int, textColor: Int, radiusDp: Int) : ReplacementSpan() {
    private var backColor = Color.GRAY
    private var textColor = Color.argb(255, 30, 30, 30)
    private var radius = App.dp2px(4)


    init {
        setBackColor(backColor)
        setTextColor(textColor)
        setRadius(radiusDp)
    }

    fun setBackColor(backColor: Int) {
        this.backColor = backColor
    }

    fun setTextColor(textColor: Int) {
        this.textColor = textColor
    }

    fun setRadius(dp: Int) {
        this.radius = App.dp2px(dp)
    }

    override fun draw(
        canvas: Canvas, text: CharSequence, start: Int, end: Int, x: Float, top: Int, y: Int,
        bottom: Int, paint: Paint
    ) {
        val rect = RectF(x, top.toFloat(), x + measureText(paint, text, start, end), bottom.toFloat())
        paint.color = backColor
        canvas.drawRoundRect(rect, radius.toFloat(), radius.toFloat(), paint)
        paint.color = textColor
        canvas.drawText(text, start, end, x, y.toFloat(), paint)
    }

    override fun getSize(paint: Paint, text: CharSequence, start: Int, end: Int, fm: Paint.FontMetricsInt?): Int {
        return Math.round(measureText(paint, text, start, end))
    }

    private fun measureText(paint: Paint, text: CharSequence, start: Int, end: Int): Float {
        return paint.measureText(text, start, end)
    }
}